### PR TITLE
feat: 💥 support tableLayout and column.ellipsis

### LINKED ISSUE #5753: Option to add ellipsis (...) to table cells, headers when content doesn't fit
[Issuehunt badges](https://issuehunt.io/r/ant-design/ant-design/issues/5753)

### What problem does this feature solve?
Right now when the table is resized and the content doesn't fit, the text is broken and reflows to the next row, thus appearing rather bad. Adding an option to allow ellipsis sign (...) by adding `text-overflow:ellipsis;` where required to appear would increase the usability IMHO.

Thanks in advance!

### What does the proposed API look like?
`ellipsize={true}` in column properties

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $25.00)

- [rororofff rororofff](https://issuehunt.io/u/rororofff) ($10.00)
- [swillis12 swillis12](https://issuehunt.io/u/swillis12) ($10.00)
- [jfbloom22 jfbloom22](https://issuehunt.io/u/jfbloom22) ($5.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/ant-design/ant-design/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### LINKED ISSUE #17127: Use table-layout fixed in scrollable or fixed-column Table
Issue: https://github.com/ant-design/ant-design/issues/17097
Reproduce: https://codesandbox.io/s/sad-cartwright-swwve
Fixed: https://codesandbox.io/s/floral-framework-e49qd

image

### PR DESCRIPTION
**Pull request by afc163**
### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

https://github.com/react-component/table/pull/369

close https://github.com/ant-design/ant-design/issues/13825#issuecomment-449889241
close #17127
close #5753

### 💡 Background and solution

1. `<Table tableLayout="fixed" />`
2. [column.ellipsis](https://deploy-preview-17284--ant-design.netlify.com/components/table-cn/#components-table-demo-ellipsis)

image

```diff
const columns = [{
   title: 'Long Column Long Column Long Column',
   dataIndex: 'address',
   key: 'address 2',
+  ellipsis: true,
}]
```

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | - Table supports `tableLayout` property for [table-layout](https://developer.mozilla.org/zh-CN/docs/Web/CSS/table-layout) attribute. And using `tableLayout="fixed"` defaully in scrollable table situation to resolve align issue caused by cell content.<br />- Table supports `column.ellipsis` to ellipsize cell content.  |
| 🇨🇳 Chinese | - Table 新增 `tableLayout` 属性，支持设置表格的 [table-layout](https://developer.mozilla.org/zh-CN/docs/Web/CSS/table-layout) 布局，并在固定表头/列下默认开启 `tableLayout="fixed"`，解决因为表格自动根据内容排版造成的列对齐问题。<br />- Table 新增 `column.ellipsis` 支持单元格内容自动省略。  |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#11822: 关于固定表头必须要给表头设置宽度](https://issuehunt.io/repos/34526884/issues/11822)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by zombieJ** [comment]
放 4.0 里？

**Comment by afc163** [comment]
这个加 3.x 里能解决当前很多问题。

**Comment by afc163** [comment]
核心改动在：react-component/table#369

**Comment by afc163** [comment]
resolve #17454 too.

**Comment by afc163** [comment]
@zombieJ need help for review.

File: components/table/demo/dynamic-settings.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ---
-order: 27
+order: 100
```
Comment (zombieJ) [inline components/table/demo/dynamic-settings.md]: 28？

**Comment by zombieJ** [comment]
Dynamic Settings 例子里可以加上 ellipsis 和 fixed 的设置

File: components/table/Table.tsx [hunk 2]
```diff
@@ -122,6 +122,7 @@ export default class Table<T> extends React.Component<TableProps<T>, TableState<
     showHeader: true,
     sortDirections: ['ascend', 'descend'],
     childrenColumnName: 'children',
+    tableLayout: 'auto',
```
Comment (zombieJ) [inline components/table/Table.tsx]: `tableLayout` 在 `rc-table` 里看起来不是一个受控属性： https://github.com/react-component/table/pull/369/files#r320658914

File: components/table/demo/dynamic-settings.md (hunk 1 above)
Comment (afc163) [reply components/table/demo/dynamic-settings.md]: 这个例子注定要放最后一个，直接改到 max 避免以后还要改。

File: components/table/Table.tsx (hunk 2 above)
Comment (afc163) [reply components/table/Table.tsx]: - https://github.com/react-component/table/commit/159e209d415229c7986e9fa1d5f594629eba25ef
- https://github.com/ant-design/ant-design/pull/17284/commits/3603942202f5d6285d084ac517f40ba2d9295d42

**Comment by afc163** [comment]
> Dynamic Settings 例子里可以加上 ellipsis 和 fixed 的设置

已加

**Comment by afc163** [comment]
用例都挂了：

image

**Comment by afc163** [comment]
原来是我把这个发掉了：https://github.com/react-component/menu/commit/72242cf528d06be78df5e63ce621ffc7ec14a417

**Comment by zombieJ** [comment]
👍 看着 OK 了，CI fix 了就可以合了。

**Review by zombieJ — APPROVED** [review]

**Comment by jeessy2** [comment]
省略能自动加title不呢？不然列表里面看不清到底是啥内容

**Comment by fengyun2** [comment]
表头能否也加上这个属性呢？

**Comment by Airkro** [comment]
在上一版本中 `columns.width` 设为 `1` 可以达到 `min-content` 的效果。但现在不管用了

我使用 tableLayout='auto' 可以解决，但该属性没有生效。

**Comment by afc163** [comment]
> 表头能否也加上这个属性呢？

@fengyun2 之前试过不太好加，可以来试试给个 PR

**Comment by Jkanon** [comment]
我加上这个属性之后，表头单元格也省略了，我感觉得表头标题不应该省略，应该加个属性切换

**Comment by HoikanChan** [comment]
antd已经上了这个功能了
https://ant.design/components/table-cn/#components-table-demo-ellipsis
麻烦加上该功能

**Comment by wuzhonghao1** [comment]
table中的columns.ellipsis属性与table的scroll属性在火狐上不兼容，出现不对齐且无法实现省略号的问题

**Pull request merged**